### PR TITLE
Implementation of the `Number` type

### PR DESCRIPTION
**Pull request by tmokenc**
# Change
The `Number` API is now looks like this
```rs
pub fn to_string(&self, radix: Radix, precision: u8) -> String;
```

Where `Radix` is an enum of { Bin, Oct, Dec, Hex } and the `precision` is maximum number of fraction part of the Number

# Unit tests
The unit tests is written directly in the documentation since it's [the Rust way of writing unit test for functions](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html), it also demonstates how the function works while reading the documentation.

Please @TheRetikGM @Blazeo7 review them for me in case I missed something that can lead into undefined behaviour like division zero, especially for those trigonometric functions.

# Tasks
- [x] Documentation 
- [x] Unit tests
- [ ] Random
- [ ] Basic arithmetic operation
    - [ ] Add
    - [ ] Sub
    - [ ] Mul
    - [ ] Div
- [ ] Pow
- [ ] Abs 
- [ ] Factorial
- [ ] Logarithm
    - [ ] log
    - [ ] log2
    - [ ] log10
    - [ ] ln 
- [ ] root
    - [ ] sqrt
- [ ] Trigonometric functions
    - [ ] sin
    - [ ] cos
    - [ ] tg
    - [ ] cotg
- [ ] Inverse trigonometric functions
    - [ ] arcsin
    - [ ] arccos
    - [ ] arctg
    - [ ] arccotg
- [ ] Combination

### CODE REVIEW

**Comment by TheRetikGM** [comment]
Will check it tomorrow.

File: src/math/src/number.rs [hunk 1]
```diff
@@ -1,18 +1,523 @@
...
+
+    /// Add two numbers together
+    ///
+    /// ```
+    /// # use math::Number;
+    ///
+    /// let a = Number::from(0.1);
+    /// let b = Number::from(0.2);
+    ///
+    /// assert_eq!(a.add(b), Ok(Number::from(0.3)));
```
Comment (TheRetikGM) [inline src/math/src/number.rs:104]: It is possible that these will fail, because:
```rust
fn main() {
    let a = 0.1 + 0.2;      // Equals to 0.30000000000000004
    let b = 0.3;            // Equals to 0.3
    if a == b {
        println!("Equal");
    } else {
        println!("NOT equal");      // Will print this
    }
}
```
I have noticed, that number `0.1` is used in other unit tests as well. Maybe we should use numbers such as `0.5` and `0.25`, which are finite and short in binary.

File: src/math/src/number.rs [hunk 2]
```diff
@@ -1,18 +1,523 @@
...
+    /// ```
+    pub fn power(&self, exp: impl Into<Self>) -> Result<Self> {
+        todo!()
+    }
+
+    /// Get the remainder of `self / other`
+    ///
+    /// ```
+    /// # use math::Number;
+    /// assert_eq!(Number::from(5).modulo(2), Ok(Number::ONE));
```
Comment (TheRetikGM) [inline src/math/src/number.rs:182]: Maybe add tests for negative numbers.

File: src/math/src/number.rs [hunk 3]
```diff
@@ -1,18 +1,523 @@
...
+    /// The number is not limited to integer, it can be a fraction
+    ///
+    /// # Error
+    /// return Error::FactorialNegative if the number is less than 0
+    ///
+    /// ```
+    /// # use math::Number;
+    /// assert_eq!(Number::Zero.factorial(), Number::One);
+    /// assert_eq!(Number::from(5).factorial(), Ok(Number::from(120)));
+    /// assert_eq!(Number::from(3.2).factorial().unwrap().to_string(6), "7.756689");
```
Comment (TheRetikGM) [inline src/math/src/number.rs:212]: `factorial(3.2)` == `7.756689535793181`, which rounded to 6 places is actually `7.756690`

File: src/math/src/number.rs [hunk 4]
```diff
@@ -1,18 +1,523 @@
...
+    /// ```
+    /// # use math::Number;
+    ///
+    /// # fn main() -> math::Result<()> {
+    ///     assert!(Number::random().log(0).is_err());
+    ///     assert!(Number::random().log(-1.2).is_err());
+    ///     assert!(Number::ZERO.log(Number::random()).is_err());
+    ///     assert!(Number::from(-0.3).log(Number::random()).is_err());
+    ///
+    ///     let a = Number::random();
```
Comment (TheRetikGM) [inline src/math/src/number.rs:234]: I am thinking, that using random floats in range <0, 1> in unit tests is not a good idea. Floating arithmetic can be weird, and thus the test could fail even when the computation is correct.

File: src/math/src/number.rs (hunk 1 above)
Comment (tmokenc) [reply src/math/src/number.rs:104]: This is actually what I want, because for those float point errors in binary representation shouldn't be in a calculator
I'm currently looking for other way to store a number instead of `f64` and that test must pass.

File: src/math/src/number.rs [hunk 5]
```diff
@@ -1,18 +1,523 @@
...
+    /// # Error
+    /// Error::ZeroNthRoot if the `nth` is 0
+    /// Error::NegativeRoot if the number is negative
+    ///
+    /// ```
+    /// # use math::Number;
+    ///
+    /// # fn main() -> math::Result<()> {
+    ///     assert!(Number::random().root(0).is_err());
+    ///     assert!(Number::from(-1).root(Number::random()).is_err());
```
Comment (TheRetikGM) [inline src/math/src/number.rs:283]: This should only fail when the root `n` is even.

File: src/math/src/number.rs (hunk 4 above)
Comment (tmokenc) [reply src/math/src/number.rs:234]: Exactly that is the point of using random <0, 1>, the test should covers cases where things easily go wrong, [I'm currently looking into this issue with binary representation of a float point.](https://github.com/FIT-IVS-segmentation-fault-core-dumped/IVS-projekt-2/pull/5#discussion_r1155293072)

File: src/math/src/number.rs (hunk 2 above)
Comment (tmokenc) [reply src/math/src/number.rs:182]: Oh yeah sure. Added

File: src/math/src/number.rs (hunk 1 above)
Comment (TheRetikGM) [reply src/math/src/number.rs:104]: Then we could use [bigdecimal](https://docs.rs/bigdecimal/latest/bigdecimal) crate for this

File: src/math/src/number.rs (hunk 3 above)
Comment (tmokenc) [reply src/math/src/number.rs]: Fixed, thanks!

File: src/math/src/number.rs (hunk 5 above)
Comment (tmokenc) [reply src/math/src/number.rs]: Fixed, thanks!

File: src/math/src/number.rs (hunk 1 above)
Comment (tmokenc) [reply src/math/src/number.rs:104]: That one has problem with performance since the number is always allocated on the heap, also it doesn't have pow function...
I'm planning to go with the [`Ratio`](https://docs.rs/num/latest/num/rational/struct.Ratio.html) type because it fits everything we need.

Comment (TheRetikGM) [reply src/math/src/number.rs]: Alright

File: src/math/src/number.rs [hunk 6]
```diff
@@ -1,18 +1,545 @@
...
+    /// # use math::Number;
+    ///
+    /// # fn main() -> math::Result<Self> {
+    ///     assert!(Number::combination(-1, Number::random()).is_err());
+    ///     assert!(Number::combination(Number::random(), -1).is_err());
+    ///
+    ///     let n = Number::random();
+    ///
+    ///     // if k > n => C(n, k) == 0
+    ///     assert_eq!(Number::combination(4, 3)?, Number::ZERO);
```
Comment (TheRetikGM) [inline src/math/src/number.rs:510]: Swapped arguments for `Number::combination`.

**Comment by TheRetikGM** [comment]
I am thinking, since the datatype for `Number` doesn't affect the rest of the app, should we merge it now and then change it after?

File: src/math/src/number.rs (hunk 1 above)
Comment (TheRetikGM) [reply src/math/src/number.rs]: However, won't there be a problem representing irrational numbers and (for example) trigonometric functions?

File: src/math/src/number.rs (hunk 6 above)
Comment (tmokenc) [reply src/math/src/number.rs:510]: Oh I didn't notice it, thanks!

**Comment by tmokenc** [comment]
Sure, the API for `Number` is considered stable with enough tests now, only need to implement them and it will be in other PR.
